cli/src/runner.rs

```diff
@@ -179,54 +202,66 @@ pub async fn run_quarantine(
     delay: std::iter::Take<ExponentialBackoff>,
 ) -> anyhow::Result<QuarantineRunResult> {
-    let quarantine_results = if run_result.failures.is_empty() {
-        QuarantineBulkTestStatus {
-            group_is_quarantined: false,
-            quarantine_results: Vec::new(),
-        }
-    } else {
+    let mut quarantine_results = QuarantineBulkTestStatus {
+        group_is_quarantined: false,
+        quarantine_results: Vec::new(),
+    };
+    let quarantine_config: QuarantineConfig = if !run_result.failures.is_empty() {
         log::info!("Quarantining failed tests");
         let result = Retry::spawn(delay, || {
-            get_quarantine_bulk_test_status(
-                api_address,
-                token,
-                org_url_slug,
-                &repo.repo,
-                &run_result.failures,
-            )
+            get_quarantining_config(api_address, token, org_url_slug, &repo.repo)
         })
         .await;
 
         result.unwrap_or_else(|e| {
             log::error!("Failed to get quarantine results: {:?}", e);
-            QuarantineBulkTestStatus {
-                group_is_quarantined: false,
-                quarantine_results: Vec::new(),
+            QuarantineConfig {
+                is_preview_mode: false,
+                quarantined_tests: Vec::new(),
             }
         })
+    } else {
+        log::info!("No failed tests to quarantine");
+        QuarantineConfig {
+            is_preview_mode: false,
+            quarantined_tests: Vec::new(),
+        }
     };
 
-    // Print out failed tests and their quarantine status
-    let quarantined_names: HashSet<_> = quarantine_results
-        .quarantine_results
-        .iter()
-        .map(|qr| &qr.name)
-        .collect();
-
+    // quarantine the failed tests
+    let quarantined: HashSet<_> = quarantine_config.quarantined_tests.iter().collect();
     for failure in &run_result.failures {
-        if quarantined_names.contains(&failure.name) {
-            log::error!("{} -> {} [QUARANTINED]", failure.parent_name, failure.name);
+        if quarantined.contains(&failure.id.as_ref().unwrap()) {
+            log::error!(
+                "{} -> {} [QUARANTINED] (id: {:?})",
+                failure.parent_name,
+                failure.name,
+                failure.id.clone().unwrap()
+            );
+
+            quarantine_results.quarantine_results.push(failure.clone());
         } else {
-            log::error!("{} -> {}", failure.parent_name, failure.name);
+            log::error!(
+                "{} -> {} (id: {:?})",
+                failure.parent_name,
+                failure.name,
+                failure.id.clone().unwrap()
+            );
         }
     }
+    quarantine_results.group_is_quarantined = run_result.failures.iter().all(|failure| {
+        quarantine_results
+            .quarantine_results
+            .iter()
+            .any(|quarantined_failure| quarantined_failure.id == failure.id)
+    });
```

Comment:
I think this logic is hard to follow and needlessly clones a lot of things. From what I can see this boils down very simply to the following:
```suggestion
    let total_failures = run_result.failures.len();
    quarantine_results.quarantine_results = run_result
        .failures
        .into_iter()
        .filter_map(|failure| {
            let quarantine_failure = quarantined.contains(&failure.id);
            log::error!(
                "{} -> {}{}(id: {})",
                failure.parent_name,
                failure.name,
                if quarantine_failure {
                    " [QUARANTINED] "
                } else {
                    " "
                },
                failure.id
            );
            if quarantine_failure {
                Some(failure)
            } else {
                None
            }
        })
        .collect();
    quarantine_results.group_is_quarantined =
        quarantine_results.quarantine_results.len() == total_failures;
```

Using `into_iter` relies on the change to accepting `run_result` as a value and not a reference. The reason for using `into_iter` instead of `iter` is to prevent the need for `clone`.